src/java.base/share/classes/java/lang/template/TemplateRuntime.java

```diff
@@ -0,0 +1,276 @@
+ * are intended for internal use only.
+ *
+ * @since 20
+ */
+@PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
+final class TemplateRuntime {
+
+    /**
+     * {@link MethodHandle} to {@link TemplateRuntime#getValue(int, StringTemplate)}.
+     */
+    private static final MethodHandle GET_VALUE_MH;
+
+    /**
+     * Initialize {@link MethodHandle MethodHandles}.
+     */
+    static {
+        try {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+
+            MethodType mt = MethodType.methodType(Object.class, int.class, StringTemplate.class);
+            GET_VALUE_MH = lookup.findStatic(TemplateRuntime.class, "getValue", mt);
+        } catch (ReflectiveOperationException | SecurityException ex) {
+            throw new AssertionError("template runtime init fail", ex);
+        }
+    }
+
+    /**
+     * Private constructor.
+     */
+    private TemplateRuntime() {
+        throw new AssertionError("private constructor");
+    }
+
+    private static final JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
+
+    private static final JavaLangInvokeAccess JLIA = SharedSecrets.getJavaLangInvokeAccess();
+
+    private static final JavaUtilCollectionAccess JUCA = SharedSecrets.getJavaUtilCollectionAccess();
+
+    /**
+     * Collect nullable elements from an array into a unmodifiable list.
+     *
+     * @param elements  elements to place in list
+     *
+     * @return unmodifiable list.
+     *
+     * @param <E>  type of elements
+     */
+    @SafeVarargs
+    @SuppressWarnings("unchecked")
+    private static <E> List<E> toList(E... elements) {
+        return JUCA.listFromTrustedArrayNullsAllowed(elements);
+    }
+
+    /**
+     * Return the types of a {@link StringTemplate StringTemplate's} values.
+     *
+     * @param st  StringTemplate to examine
+     *
+     * @return list of value types
+     *
+     * @throws NullPointerException if st is null
+     *
+     * @implNote The default method determines if the {@link StringTemplate}
+     * was synthesized by the compiler, then the types are precisely those of the
+     * embedded expressions, otherwise this method returns the values list types.
+     */
+    static List<Class<?>> valueTypes(StringTemplate st) {
+        Objects.requireNonNull(st, "st must not be null");
+        List<Class<?>> result = new ArrayList<>();
+        Class<?> tsClass = st.getClass();
+        if (tsClass.isSynthetic()) {
+            try {
+                for (int i = 0; ; i++) {
+                    Field field = tsClass.getDeclaredField("x" + i);
+                    result.add(field.getType());
+                }
+            } catch (NoSuchFieldException ex) {
+                // End of fields
+            } catch (SecurityException ex) {
+                throw new InternalError(ex);
+            }
+
+            return result;
+        }
+        for (Object value : st.values()) {
```

Comment:
I think that valueTypes() should return the types of the values not the dynamic classes of the values.
Here there is no type information so it should be Object for all values.

It has also the nice property that the return type of the accessors returned by valueAccessors are the same as valueTypes() which is i believe more coherent.

Reply:
Changing